collect_app/src/main/java/org/odk/collect/android/fragments/ShowQRCodeFragment.java

```diff
@@ -0,0 +1,316 @@
+import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.odk.collect.android.R;
+import org.odk.collect.android.preferences.PreferenceKeys;
+import org.odk.collect.android.utilities.TextUtils;
+import org.odk.collect.android.utilities.ToastUtils;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.zip.DataFormatException;
+
+import timber.log.Timber;
+
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_FORMLIST_URL;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_GOOGLE_SHEETS_URL;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_PASSWORD;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_PROTOCOL;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_SERVER_URL;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_SUBMISSION_URL;
+import static org.odk.collect.android.preferences.PreferenceKeys.KEY_USERNAME;
+
+
+/**
+ * Created by shobhit on 6/4/17.
+ */
+
+public class ShowQRCodeFragment extends Fragment implements View.OnClickListener {
+
+    private static final int SELECT_PHOTO = 111;
+    private SharedPreferences settings;
+    private ShareActionProvider mShareActionProvider;
+    private ImageView qrImageView;
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        return inflater.inflate(R.layout.show_qrcode_fragment, container, false);
+    }
+
+    @Override
+    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+
+        settings = PreferenceManager.getDefaultSharedPreferences(getActivity());
+        setHasOptionsMenu(true);
+
+        qrImageView = (ImageView) view.findViewById(R.id.qr_iv);
+        Button scan = (Button) view.findViewById(R.id.btnScan);
+        scan.setOnClickListener(this);
+        Button select = (Button) view.findViewById(R.id.btnSelect);
+        select.setOnClickListener(this);
+    }
+
+    public void generateCode() {
+        Bitmap qrCode = generateQRBitMap();
+        if (qrCode != null) {
+            qrImageView.setImageBitmap(qrCode);
+            updateShareIntent(qrCode);
+        }
+    }
+
+    private void updateShareIntent(Bitmap qrCode) {
+
+        //Save the bitmap to a file
+        File cache = getActivity().getApplicationContext().getExternalCacheDir();
+        File shareFile = new File(cache, "shareImage.jpeg");
+        try {
+            FileOutputStream out = new FileOutputStream(shareFile);
+            qrCode.compress(Bitmap.CompressFormat.JPEG, 100, out);
+            out.flush();
+            out.close();
```

Comment:
I think you don't need to call flush and move the call to close in finally block?